pylib/anki/collection.py

```diff
@@ -696,6 +696,16 @@ def _pb_search_separator(self, operator: SearchJoiner) -> SearchNode.Group.Joine
     # Browser Table
     ##########################################################################
 
+    def all_browser_card_columns(
+        self,
+    ) -> Generator[Tuple[str, str, bool, bool, bool, bool], None, None]:
```

Comment:
The unlabeled tuple is not a very nice signature for a public API. I'm guessing the motivation for this was that the dataclass makes reference to add-ons, and you thought that didn't belong in pylib? Since this is just a "data transfer object" that doesn't need any special logic associated with it, what I'd recommend you do instead is make the protobuf object public by adding it to the top of collection.py, and then return that protobuf message instead. We could then either convert directly from that protobuf to the dataclass, or just use it as-is. We'll need to run it through a hook for add-ons to inject the correct label for a given key anyway, and at display time we could just do "column.label or tr.browsing_addon()" - so I'd lean towards dropping the dataclass unless you think there's some other reason we'd need it.

Reply:
I rather thought we would add a hook after the state was set up. Add-ons could then modify the column dict and the active column list directly.
My main reason to avoid the pb class was it's opaqueness. We don't have auto completion or type hints and even finding its definition is tough for people unfamiliar with the source code. It's the same approach I used for CellRow but I see that Column is a pure dataclass in contrast. Also, if we use enums instead of bools as you suggested below, it makes more sense to use the pb structures directly.

Reply:
The symlinks in the source tree are to help editors like PyCharm/intelliJ pick up the generated types:

```
dae@dip:pylib/anki/_backend% ls -l
total 56
-rw-r--r--  1 dae  staff  2019 Apr  3 10:25 BUILD.bazel
-rw-r--r--  1 dae  staff  6819 Apr  5 11:11 __init__.py
lrwxr-xr-x  1 dae  staff    54 Feb  7 13:54 backend_pb2.pyi -> ../../../bazel-bin/pylib/anki/_backend/backend_pb2.pyi
lrwxr-xr-x  1 dae  staff    48 Apr  3 10:25 fluent.py -> ../../../bazel-bin/pylib/anki/_backend/fluent.py
-rwxr-xr-x  1 dae  staff  5382 Apr  3 10:25 genbackend.py
lrwxr-xr-x  1 dae  staff    51 Feb  7 13:54 generated.py -> ../../../bazel-bin/pylib/anki/_backend/generated.py
-rw-r--r--  1 dae  staff  2533 Apr  3 10:25 genfluent.py
-rw-r--r--  1 dae  staff   235 Apr  3 10:25 rsbridge.pyi
```

On my machine, I run ./run to ensure the build products are in bazel-bin, and then get completions/help in the editor:
<img width="1075" alt="Screen Shot 2021-04-08 at 9 42 16 am" src="https://user-images.githubusercontent.com/180542/113947798-ce3e6a00-984e-11eb-9a60-4c9ee00015e2.png">

The editor sometimes doesn't seem to notice updates, but I've found opening the .pyi file or generated.py causes it to rebuild the index.

The .pyi files are included in the Python wheels, so the typings should be available to end users as well, with the ability to click on a definition to jump to it.

If the symlinks are not working on Windows, you could write a little script to copy the files into the tree instead of symlinking them in perhaps?

Re the placement of hooks, I haven't been following that part closely - we'll need a hook somewhere, but it doesn't have to be here. :-)

Reply:
Thanks for the instructions. Now it's only a matter of me bucking up and setting up my Python environment correctly. 😅